File: app/src/main/java/com/tomclaw/appsend/util/PackageHelper.java
package com.tomclaw.appsend.util;

import android.content.pm.ApplicationInfo;
import android.content.pm.PackageManager;
import android.graphics.Bitmap;
import android.graphics.Canvas;
import android.graphics.drawable.BitmapDrawable;
import android.graphics.drawable.Drawable;

import com.tomclaw.appsend.main.item.StoreItem;

import java.io.ByteArrayOutputStream;
import java.io.IOException;

/**
 * Created by solkin on 23.04.17.
 */
public class PackageHelper {

    public static int getInstalledVersionCode(String packageName, PackageManager packageManager) {
        try {
            return packageManager.getPackageInfo(packageName, 0).versionCode;
        } catch (Throwable ex) {
            return StoreItem.NOT_INSTALLED;
        }
    }

    public static byte[] getPackageIconPng(ApplicationInfo info,
                                           PackageManager packageManager) {
        Drawable icon = info.loadIcon(packageManager);
        final Bitmap bitmap = drawableToBitmap(icon);
        ByteArrayOutputStream baos = new ByteArrayOutputStream();
        bitmap.compress(Bitmap.CompressFormat.PNG, 100, baos);
        byte[] data = baos.toByteArray();
        try {
            baos.close();
        } catch (IOException ignored) {
        }
        return data;
    }

    private static Bitmap drawableToBitmap(Drawable drawable) {
        Bitmap bitmap;

        if (drawable instanceof BitmapDrawable) {
            BitmapDrawable bitmapDrawable = (BitmapDrawable) drawable;
            if (bitmapDrawable.getBitmap() != null) {
                return bitmapDrawable.getBitmap();
            }
        }

        if (drawable.getIntrinsicWidth() <= 0 || drawable.getIntrinsicHeight() <= 0) {
            bitmap = Bitmap.createBitmap(1, 1, Bitmap.Config.ARGB_8888);
        } else {
            bitmap = Bitmap.createBitmap(drawable.getIntrinsicWidth(), drawable.getIntrinsicHeight(), Bitmap.Config.ARGB_8888);
        }

        Canvas canvas = new Canvas(bitmap);
        drawable.setBounds(0, 0, canvas.getWidth(), canvas.getHeight());
        drawable.draw(canvas);
        return bitmap;
    }
}

File: app/src/main/java/com/tomclaw/appsend/util/PackageHelper.kt
package com.tomclaw.appsend.util

import android.content.Context
import android.content.pm.ApplicationInfo
import android.content.pm.PackageManager
import android.graphics.Bitmap
import android.graphics.Canvas
import android.graphics.drawable.BitmapDrawable
import android.graphics.drawable.Drawable
import androidx.appcompat.content.res.AppCompatResources
import androidx.core.graphics.createBitmap
import com.tomclaw.appsend.R
import java.io.ByteArrayOutputStream

object PackageHelper {

    fun getPackageIconPng(
        info: ApplicationInfo?,
        packageManager: PackageManager,
        context: Context
    ): ByteArray {
        val icon = info?.loadIcon(packageManager)
            ?: AppCompatResources.getDrawable(context, R.drawable.app_placeholder)
            ?: return ByteArray(0)
        val bitmap = drawableToBitmap(icon)
        return ByteArrayOutputStream().use { baos ->
            bitmap.compress(Bitmap.CompressFormat.PNG, 100, baos)
            baos.toByteArray()
        }
    }

    private fun drawableToBitmap(drawable: Drawable): Bitmap {
        if (drawable is BitmapDrawable && drawable.bitmap != null) {
            return drawable.bitmap
        }

        val bitmap = if (drawable.intrinsicWidth <= 0 || drawable.intrinsicHeight <= 0) {
            createBitmap(width = 1, height = 1)
        } else {
            createBitmap(drawable.intrinsicWidth, drawable.intrinsicHeight)
        }

        val canvas = Canvas(bitmap)
        drawable.setBounds(0, 0, canvas.width, canvas.height)
        drawable.draw(canvas)
        return bitmap
    }

}
